miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+            offset += bits
+        if offset > self._num.size() * 8:
+            raise ValueError("sum of bit lengths is > to the backing num size")
+        super(BitField, self).__init__(fields)
+
+    def set(self, vm, addr, val):
+        self._num.set(vm, addr, val)
+
+    def get(self, vm, addr):
+        return self._num.get(vm, addr)
+
+    def __eq__(self, other):
+        return self.__class__ == other.__class__ and \
+                self._num == other._num and super(BitField, self).__eq__(other)
+
+    def __hash__(self):
+        return hash(super(BitField, self).__hash__() + hash(self._num))
+
+
+# MemStruct classes
+
+class _MetaMemStruct(type):
+    """MemStruct metaclass. Triggers the magic that generates the class fields
+    from the cls.fields list.
+
+    Just calls MemStruct.gen_fields(), the actual implementation can seen be
+    there.
+    """
+
+    def __init__(cls, name, bases, dct):
+        super(_MetaMemStruct, cls).__init__(name, bases, dct)
+        cls.gen_fields()
+
+    def __repr__(cls):
+        return cls.__name__
+
+
+class MemStruct(object):
+    """Base class to implement VmMngr backed C-like structures in miasm.
+
+    The mechanism is the following:
+        - set a "fields" class field to be a list of
+          (<field_name (str)>, <MemField_subclass_instance>)
+        - instances of this class will have properties to interract with these
+          fields.
+
+    Example:
+        class Example(MemStruct):
+            fields = [
+                # Number field: just struct.pack fields with one value
+                ("num", Num("I")),
+                ("flags", Num("B")),
+                # Ptr fields are Num, but they can also be dereferenced
+                # (self.deref_<field>). Deref can be read and set.
+                ("other", Ptr("I", OtherStruct)),
+                ("i", Ptr("I", Num("I"))),
+                # Ptr to a variable length String
+                ("s", Ptr("I", MemStr)),
+            ]
+
+        mstruct = MyStruct(vm, addr)
+
+        # Field assignment modifies virtual memory
+        mstruct.num = 3
+        assert mstruct.num == 3
+        memval = struct.unpack("I", vm.get_mem(mstruct.get_addr(),
+                                                      4))[0]
+        assert memval == mstruct.num
+
+        # Memset sets the whole structure
+        mstruct.memset()
+        assert mstruct.num == 0
+        mstruct.memset('\x11')
+        assert mstruct.num == 0x11111111
+
+        other = OtherStruct(vm, addr2)
+        mstruct.other = other.get_addr()
+        assert mstruct.other == other.get_addr()
+        assert mstruct.deref_other == other
+        assert mstruct.deref_other.foo == 0x1234
+
+    See the various MemField doc for more information.
+    """
+    __metaclass__ = _MetaMemStruct
+
+    fields = []
+
+    _size = None
+
+    # Classic usage methods
+
+    def __init__(self, vm, addr=None, *args, **kwargs):
+        global ALLOCATOR
```

Comment:
The ALLOCATOR could be a class attribute, rewritable with an instance attribute. That way, different allocators can be used in the same run (for instance, with multiple heap in Windows).